pkg/operator/controller.go

```diff
@@ -3,482 +3,293 @@ package operator
-		workqueue:  workqueue.NewRateLimitingQueue(workqueue.DefaultControllerRateLimiter()),
-		listers:    listers,
-		clients:    clients,
-		pc:         NewProfileCalculator(listers, clients),
-	}
+// SetupWithManager creates a new Tuned Controller and adds it to the Manager.
+// The Manager will set fields on the Controller and start it when the Manager is started.
+func (r *TunedReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	profilePredicates := predicate.Funcs{
+		UpdateFunc: func(e event.UpdateEvent) bool {
+			if !r.validateUpdateEvent(&e) {
+				return false
+			}
 
-	// Initial event to bootstrap CR if it doesn't exist.
-	controller.workqueue.AddRateLimited(wqKey{kind: wqKindTuned, name: tunedv1.TunedDefaultResourceName})
+			profileOld := e.ObjectOld.(*tunedv1.Profile)
+			profileNew := e.ObjectNew.(*tunedv1.Profile)
 
-	controller.clients.Kube, err = kubeset.NewForConfig(controller.kubeconfig)
-	if err != nil {
-		return nil, err
+			return !reflect.DeepEqual(profileOld.Status.Conditions, profileNew.Status.Conditions)
+		},
 	}
 
-	// ClusterOperator
-	controller.clients.ConfigV1Client, err = configv1client.NewForConfig(controller.kubeconfig)
-	if err != nil {
-		return nil, err
+	podPredicates := predicate.Funcs{
+		CreateFunc: func(e event.CreateEvent) bool {
+			return r.state.podEventsEnabled
+		},
+		DeleteFunc: func(e event.DeleteEvent) bool {
+			return r.state.podEventsEnabled
+		},
 	}
 
-	// Tuned
-	controller.clients.Tuned, err = tunedset.NewForConfig(controller.kubeconfig)
-	if err != nil {
-		return nil, err
+	mcpPredicates := predicate.Or(
+		predicate.GenerationChangedPredicate{},
+		predicate.LabelChangedPredicate{},
+		predicate.Funcs{UpdateFunc: func(e event.UpdateEvent) bool {
+			if !validateUpdateEvent(&e) {
+				return false
+			}
+			mcpOld := e.ObjectOld.(*mcov1.MachineConfigPool)
+			mcpNew := e.ObjectNew.(*mcov1.MachineConfigPool)
+			return !reflect.DeepEqual(mcpOld.Spec, mcpNew.Spec)
+		}})
+
+	// Field indexer is required for listing pods matching the spec.nodeName field.
+	// The alternative is to use an uncached client.
+	if err := mgr.GetFieldIndexer().IndexField(context.TODO(), &corev1.Pod{}, "spec.nodeName", func(o client.Object) []string {
+		pod := o.(*corev1.Pod)
+		return []string{pod.Spec.NodeName}
+	}); err != nil {
+		return err
 	}
 
-	// MachineConfig
-	controller.clients.MC, err = mcfgclientset.NewForConfig(controller.kubeconfig)
+	err := ctrl.NewControllerManagedBy(mgr).
+		For(&tunedv1.Tuned{}).
+		Owns(&tunedv1.Profile{}, builder.WithPredicates(profilePredicates)).
+		Watches(
+			&source.Kind{Type: &corev1.Node{}},
+			handler.EnqueueRequestsFromMapFunc(r.defaultTunedRequests),
+			//TODO - consider using builder.OnlyMetadata instead these perdicats
+			builder.WithPredicates(predicate.LabelChangedPredicate{})).
+		Watches(
+			&source.Kind{Type: &corev1.Pod{}},
+			handler.EnqueueRequestsFromMapFunc(r.podLabelsToTuned),
+			builder.WithPredicates(podPredicates,
```

Comment:
I did a little bit of testing when creating 1000 pods with the operator config when pod-matching functionality is not in use and the impact on the CPU is noticeable.  This is a clear regression over the old implementation and is likely to be picked up when the Perf team does large-scale OCP testing.  Is there another way to implement this in controller-runtime to make sure we do not regress at all?

Reply:
I was looking into https://github.com/kubernetes-sigs/controller-runtime/issues/246  which will allow us to add a watch for pods dynamically but I still need to figure out how to remove the watch (if this is even possible).